onnxruntime/core/providers/cpu/nn/string_split.cc

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cpu/nn/string_split.h"
+#include <algorithm>
+#include <limits>
+#include <string>
+#include "core/common/common.h"
+namespace onnxruntime {
+
+ONNX_CPU_OPERATOR_KERNEL(StringSplit, 20,
+                         KernelDefBuilder()
+                             .TypeConstraint("T1", DataTypeImpl::GetTensorType<std::string>())
+                             .TypeConstraint("T2", DataTypeImpl::GetTensorType<std::string>())
+                             .TypeConstraint("T3", DataTypeImpl::GetTensorType<int64_t>()),
+                         StringSplit);
+
+/// Calculate substrings in ``str`` delimited by ``delimiter``. A maximum of ``max_splits`` splits are permitted.
+/// Returns a vector of string slices into ``str`` representing the substrings as string views. The user must ensure
+/// the returned views' lifetime does not exceed ``str``'s.
+InlinedVector<std::string_view> ComputeSubstrings(std::string_view str, std::string_view delimiter, int64_t max_splits) {
+  InlinedVector<std::string_view> output;
+  if (str.empty()) {
+    return output;
+  }
+  if (delimiter.empty()) {
+    // Count consecutive whitespace as one delimiter. Preceding and trailing whitespace is meant to be ignored.
+    size_t pos = str.find_first_not_of(" ");
+    int64_t token_count = 0;
+    while (pos != std::string::npos) {
+      if (token_count++ == max_splits) {
+        // Trim down last substring as required in specification
+        size_t next_pos = str.length() - 1;
+        while (str[next_pos] == ' ') {
+          next_pos--;
+        }
+        output.push_back(str.substr(pos, next_pos - pos + 1));
```

Comment:
>str.substr [](http://example.com/codeflow?start=25&length=10)

We want to use string_view to minimize data copying and point to the original string. There are two issues with using `substr()`

1) It creates a copy of the original data which defeats the purpose of the code.
2) It creates a string_view instance off that temporary substring instance which is then promptly destroyed leaving string_view pointing to  a non-existing buffer.

Reply:
Not sure what I am missing here but since str is a string_view, surely str.substr(pos, next_pos - pos + 1) should not involve copying the original data (following [cppreference](https://en.cppreference.com/w/cpp/string/basic_string_view/substr) which suggests it is a constant time operation)? Also str's lifetime is until the end of the compute method so I'm surprised it would point to an invalid buffer.

Reply:
Yes, you are correct. I missed it.